File: 프로그래머스/lv1/42889. 실패율/실패율.js


function solution(N, stages) {
    const fail = new Map();
    for (let i = 1; i <= N; i++) {
        let count = stages.filter((x) => x === i).length;
        let players = stages.filter((x) => x >= i).length;
        let failRate = 0;
        if (players !== 0) {
            failRate = count / players;
        }
        fail.set(i, failRate);
    }
    console.log(fail.entries())
    let result = [...fail.entries()].map(([stage, rate]) => ({ stage, rate }));
    result.sort((a, b) => {
        if (a.rate > b.rate) return -1;
        if (a.rate < b.rate) return 1;
        return a.stage - b.stage;
    });
    return result.map((x) => x.stage);
}

File: 프로그래머스/1/42889. 실패율/실패율.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.stream.Collectors;

class Solution {
    public int[] solution(int N, int[] stages) {
        HashMap<Integer, Double> map = new HashMap<>();

        for (int i = 1; i <= N; i++) {
            int currentStage = i;
            long count = Arrays.stream(stages).filter(stage -> stage == currentStage).count();
            long player = Arrays.stream(stages).filter(stage -> stage >= currentStage).count();
            map.put(i, (double) count / player);
        }

        List<HashMap.Entry<Integer, Double>> list;
        list = map.entrySet()
            .stream()
            .sorted((a, b) -> {
                if (a.getValue() > b.getValue()) return -1;
                if (a.getValue() < b.getValue()) return 1;
                return a.getKey().compareTo(b.getKey());
            })
            .collect(Collectors.toList());

        return list.stream().mapToInt(HashMap.Entry::getKey).toArray();

    }
}